fix(HMS-2519): error handling of jobs

lzap opened the pull request:
Looks like our error handling of the last step was not right. In some cases, reservations are kept in "pending" state when the instance info could not be retrieved. This fixes it.

Also moves handling of notifications into a helper function, we only have jobs for reservations so it is fine for now to have it here.

Finally, adds open telemetry to AWS and GCP similarly as it is already implemented for Azure.

ezr-ondrej reviewed (approved):
Nice! :)
Except the Import ordering, do you have different imports ordering set up in goland then the linter wants maybe? :thinking:

avitova commented on .golangci.yml:
```diff
@@ -20,7 +20,6 @@ linters:
     - gci # too strict
     - contextcheck # too restrictive
     - depguard # too verbose
-    - tagalign # cleanenv is unreadable
```
Happened to me. Check your golangci-lint version.
Tagalign is available from v1.53.0.

lzap replied:
Ah I had an old version of the linter locally :-D

avitova commented on internal/jobs/launch_instance_azure.go:
```diff
@@ -95,6 +91,7 @@ func DoEnsureAzureResourceGroup(ctx context.Context, args *LaunchInstanceAzureTa
 
 	azureClient, err := clients.GetAzureClient(ctx, args.Subscription)
 	if err != nil {
+		logger.Error().Err(err).Msg("cannot create new Azure client")
```
```suggestion
		logger.Error().Err(err).Msg("Cannot create new Azure client")
```

lzap replied:
Lesson learned for my next project - all logging start with lower case :-)

avitova commented on internal/jobs/launch_instance_gcp.go:
```diff
@@ -150,6 +159,9 @@ func DoLaunchInstanceGCP(ctx context.Context, args *LaunchInstanceGCPTaskArgs) e
 }
 
 func FetchInstancesDescriptionGCP(ctx context.Context, args *LaunchInstanceGCPTaskArgs) error {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "FetchInstancesDescriptionAWS")
```
```suggestion
	ctx, span := otel.Tracer(TraceName).Start(ctx, "FetchInstancesDescriptionGCP")
```

avitova commented on internal/jobs/launch_instance_gcp.go:
```diff
@@ -56,26 +58,27 @@ func HandleLaunchInstanceGCP(ctx context.Context, job *worker.Job) {
 
 // DoLaunchInstanceGCP is a job logic, when error is returned the job status is updated accordingly
 func DoLaunchInstanceGCP(ctx context.Context, args *LaunchInstanceGCPTaskArgs) error {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "FetchInstancesDescriptionAWS")
```
```suggestion
	ctx, span := otel.Tracer(TraceName).Start(ctx, "FetchInstancesDescriptionGCP")
```

avitova commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -273,13 +297,15 @@ func FetchInstancesDescriptionAWS(ctx context.Context, args *LaunchInstanceAWSTa
 			errRetry := rDao.UpdateReservationInstance(ctx, args.ReservationID, instance)
 			if errRetry != nil {
+				span.SetStatus(codes.Error, "cannot update instance description")
 				return fmt.Errorf("cannot update instance description: %w", err)
```
```suggestion
				return fmt.Errorf("cannot update instance description: %w", errRetry)
```

lzap commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -60,33 +62,32 @@ func HandleLaunchInstanceAWS(ctx context.Context, job *worker.Job) {
-	} else {
-		nc.SuccessfulLaunch(ctx, args.ReservationID)
+		finishWithError(ctx, args.ReservationID, jobErr)
 	}
```
This is the actual bug - instead of calling `finishWithError` we just sent a notification and skipped on the error.

ezr-ondrej replied:
hmmm when thinking about it, shall we fail here? as this step is not required... and we will report error on it, Users might get confused :thinking:

avitova reviewed (commented):
Some more things. Sorry for not noticing earlier. They are more or less only nits.

lzap commented:
Great finds.

I also found one suspicions code that I want to remove, we only finish a reservation in case all steps were processed. But the `finishJob` function is always called at the very end of each job. So I think this was a mistake and I remove that. I added a error log in case a job that did not run all jobs is finished.

Finally, I also added panic handing via `recover`, in case there was a panic (nil pointer) the reservation is immediately finished with failure flag and the panic value (string) is logged and stored in status.

lzap commented on internal/jobs/common.go:
```diff
@@ -31,25 +38,26 @@ func finishWithSuccess(ctx context.Context, reservationId int64) {
-		err = rDao.FinishWithSuccess(ctx, reservationId)
-		if err != nil {
-			logger.Warn().Err(err).Msg("unable to update job status: finish")
-		}
```
This smells, I do not remember why I wrote this, finish function should be only called as the very last step therefore it does not matter at which step we are. I removed this and now it simply finishes properly, tho, there is some extra logging above just to inform us in case we finish a reservation that did not execute all steps.

ezr-ondrej commented on internal/jobs/common.go:
```diff
@@ -15,11 +17,16 @@ import (
 const TraceName = telemetry.TracePrefix + "internal/jobs"
 
 var ErrTypeAssertion = errors.New("type assert error")
+var ErrPanicInJob = errors.New("panic during job")
```
No idea what linter is complaining about :thinking: 
My guess would be that it wants:
```suggestion
var (
  ErrTypeAssertion = errors.New("type assert error")
  ErrPanicInJob = errors.New("panic during job")
)
```

ezr-ondrej replied:
Yup, now it's happy! :tada:

adiabramovitch commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -100,11 +108,13 @@ func DoEnsurePubkeyOnAWS(ctx context.Context, args *LaunchInstanceAWSTaskArgs) e
 
 	pubkey, err := pkDao.GetById(ctx, args.PubkeyID)
 	if err != nil {
+		span.SetStatus(codes.Error, "cannot upload aws key")
```
```suggestion
		span.SetStatus(codes.Error, "cannot get key")
```

lzap replied:
Okay rebased

lzap commented on internal/jobs/common.go:
```diff
@@ -31,25 +38,26 @@ func finishWithSuccess(ctx context.Context, reservationId int64) {
+	if reservation.Step == reservation.Steps {
+		logger.Info().Msgf("Finishing reservation with success at step %d/%d", reservation.Step, reservation.Steps)
+	} else {
+		logger.Error().Msgf("Finishing reservation with success at step %d/%d", reservation.Step, reservation.Steps)
+	}
```
Feels weird but I want this to be logged into Glitchtip because a reservation was finished while expected amount of steps was not fulfilled. Could be a warning but we do not send warnings to Glitchtip.

lzap commented:
Fixed linter.

adiabramovitch commented:
/retest

avitova reviewed (approved):
LGTM :+1:
